python/docs/events.md

```diff
@@ -0,0 +1,204 @@
+# Emitter Events API
+
+## Event APIs
+
+Built-in events will always return a `dict[str, Any] | None`
+
+### Agent Events
+
+These events can be observed calling `agent.run`
+
+- "start":
+    ```python
+    {
+        "meta": BeeMeta,
+        "tools": list[Tool],
+        "memory": BaseMemory,
+    }
+
+- "error":
+    ```python
+    {
+        "error": FrameworkError,
+        "meta": BeeMeta,
+    }
+
+- "retry":
+    ```python
+    {
+        "meta": BeeMeta,
+    }
+
+- "success":
+    ```python
+    {
+        "data": Message,
+        "iterations": list[BeeAgentRunIteration],
+        "memory": BaseMemory,
+        "meta": BeeMeta,
+    }
+
+- "update" and "partialUpdate":
+    ```python
+    {
+        "data": BeeIterationResult | dict[str, Any],
+        "update": {
+            "key": str,
+            "value": Any,
+            "parsedValue": Any,
+        },
+        "meta": { "success": bool },
+        "memory": BaseMemory,
+        "tools": list[Tool] | None,
+    }
+
+- "toolStart":
+    ```python
+    {
+        "data": {
+            "tool": Tool,
+            "input": Any,
+            "options": BeeRunOptions,
+            "iteration": BeeIterationResult,
+        },
+        "meta": BeeMeta,
+    }
+
+- "toolSuccess":
+    ```python
+    {
+        "data": {
+            "tool": Tool,
+            "input": Any,
+            "options": BeeRunOptions,
+            "iteration": BeeIterationResult,
+            "result": ToolOutput,
+        },
+        "meta": BeeMeta,
+    }
+
+- "toolError":
+    ```python
+    {
+        "data": {
+            "tool": Tool,
+            "input": Any,
+            "options": BeeRunOptions,
+            "iteration": BeeIterationResult,
+            "error": FrameworkError,
+        },
+        "meta": BeeMeta,
+    }
+
+### ChatModel Events
+
+These events can be observed when calling `ChatModel.create` or `ChatModel.create_structure`
+
+- "newToken":
+    ```python
+    {
+      "value": ChatModelOutput,
+      "abort": Callable[[], None],
+    }
+
+- "success":
+    ```python
+    {
+      "value": ChatModelOutput
+    }
+- "start": `ChatModelInput`
+    ```python
+    {
+      "input": ChatModelInput
+    }
+- "error":
+    ```python
+    {
+      "error": ChatModelError
+    }
+- "finish": `None`
+
+### Tool Events
+
+These events can be observed when calling `Tool.run`
```

Comment:
`.run` are internal events, I would remove them

Reply:
We call `Tool.run` directly in many examples though, wouldn't that make it public?